org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/LSPTextHover.java

```diff
@@ -145,42 +145,69 @@ public class LSPTextHover implements ITextHover, ITextHoverExtension {
+
+		final IDocument document = textViewer.getDocument();
+		if (document == null) {
+			return null;
+		}
+
 		try {
-			final IDocument document = textViewer.getDocument();
-			if (document == null) {
-				return null;
-			}
 			final var oneHoverAtLeast = new boolean[] { false };
 			final var regionStartOffset = new int[] { 0 };
 			final var regionEndOffset = new int[] { document.getLength() };
-			castNonNull(this.request).get(GET_TIMEOUT_MS, TimeUnit.MILLISECONDS).stream()
-				.filter(Objects::nonNull)
-				.map(Hover::getRange)
-				.filter(Objects::nonNull)
-				.forEach(range -> {
-					try {
+			// Wait shortly for hover region result, fallback to heuristics if LS is laggy
+			castNonNull(this.request).get(100, TimeUnit.MILLISECONDS).stream() //
+					.filter(Objects::nonNull) //
+					.map(Hover::getRange) //
+					.filter(Objects::nonNull) //
+					.forEach(range -> {
+						try {
 							regionStartOffset[0] = Math.max(regionStartOffset[0],
 									LSPEclipseUtils.toOffset(range.getStart(), document));
 							regionEndOffset[0] = Math.min(regionEndOffset[0],
 									LSPEclipseUtils.toOffset(range.getEnd(), document));
-						oneHoverAtLeast[0] = true;
-					} catch (BadLocationException e) {
-						LanguageServerPlugin.logError(e);
-					}
-				});
+							oneHoverAtLeast[0] = true;
+						} catch (BadLocationException e) {
+							LanguageServerPlugin.logError(e);
+						}
+					});
 			if (oneHoverAtLeast[0]) {
 				this.lastRegion = new Region(regionStartOffset[0], regionEndOffset[0] - regionStartOffset[0]);
 				return this.lastRegion;
 			}
-		} catch (ExecutionException e) {
+		} catch (Exception e) {
 			LanguageServerPlugin.logError(e);
-		} catch (InterruptedException e) {
-			Thread.currentThread().interrupt();
-		} catch (TimeoutException e) {
-			LanguageServerPlugin.logWarning("Could not get hover region due to timeout after " + GET_TIMEOUT_MS + " milliseconds"); //$NON-NLS-1$ //$NON-NLS-2$
 		}
-		this.lastRegion = new Region(offset, 0);
-		return this.lastRegion;
+
+		// Fallback to heuristic region without blocking.
+		System.err.println("FALLBACK"); //$NON-NLS-1$
+		final Region heuristic = computeHeuristicRegion(document, offset);
+		this.lastRegion = heuristic;
+		return heuristic;
+	}
+
+	private static Region computeHeuristicRegion(final IDocument document, final int offset) {
+		try {
+			final int length = document.getLength();
+			if (offset < 0 || offset > length) {
+				return new Region(Math.max(0, Math.min(offset, length)), 0);
+			}
+			int start = offset;
+			int end = offset;
+			while (start > 0 && isWordPart(document.getChar(start - 1))) {
+				start--;
+			}
+			while (end < length && isWordPart(document.getChar(end))) {
+				end++;
+			}
+			return new Region(start, Math.max(0, end - start));
+		} catch (final BadLocationException ex) {
+			return new Region(offset, 0);
+		}
+	}
+
+	private static boolean isWordPart(char c) {
+		return Character.isLetterOrDigit(c) || c == '_' || c == '$';
```

Comment:
I know there is an extension point in eclipse that is used to calculate word boundaries on double click. Could we use that one?